File: leetcode_problems/cpp/lc1079_letter_tile_possibilities.cpp
#include <string>
using std::string;

class Solution
{
public:
    void backtrack(int arr[], int &count)
    {
        for (int i = 0; i < 26; i++)
        {
            if (arr[i] <= 0)
            {
                continue;
            }

            arr[i]--;
            count++;
            backtrack(arr, count);
            arr[i]++;
        }
    }

    int numTilePossibilities(string tiles)
    {
        int count = 0;
        int *arr = new int[26]();
        for (char chr : tiles)
        {
            arr[chr - 'A']++;
        }
        backtrack(arr, count);
        return count;
    }
};

File: leetcode_problems/typescript/src/lc1079_letter_tile_possibilities.ts
function numTilePossibilities(tiles: string): number {
  function backtrack() {
    for (let i = 0; i < 26; i++) {
      if (arr[i] <= 0) {
        continue;
      }
      arr[i]--;
      count++;
      backtrack();
      arr[i]++;
    }
  }

  let count = 0;
  let arr = new Array(26).fill(0);
  for (let chr of tiles) {
    arr[chr.charCodeAt(0) - "a".charCodeAt(0)]++;
  }
  backtrack();
  return count;
}
